pymc3/distributions/distribution.py

```diff
@@ -295,10 +296,27 @@ def draw_values(params, point=None):
                               if node is not None and
                               node.name not in stored and
                               node not in params])
-    values = []
-    for param in params:
-        values.append(_draw_value(param, point=point, givens=givens.values()))
-    return values
+
+    # Funny dance to resolve missing nodes
+    params = dict(enumerate(params))  # some nodes are not hashable
+    evaluated = {}
+    to_eval = set()
+    missing_inputs = set(params)
+    while to_eval or missing_inputs:
+        if to_eval == missing_inputs:
+            raise ValueError('Cannot resolve inputs for {}'.format([str(j) for j in to_eval]))
+        to_eval = set(missing_inputs)
+        missing_inputs = set()
+        for param in to_eval:
+            try:  # might evaluate in a bad order,
+                evaluated[param] = _draw_value(params[param], point=point, givens=givens.values(), size=size)
+                if any(param in j for j in named_nodes_children.values()):
```

Comment:
Very nice but I'm a bit confused by this line. You're trying to see if the node `param` (or maybe `params[param]`) is a child of some other named node. If `params[param]` were to be a named node, that information should be available in the dictionary `named_nodes_parents`. If `params[param]` were not to be a named node, then it would not be registered neither in the `named_nodes_parents` nor the  `named_nodes_children` dictionaries.
If `params[param]` is a named node, you should be able to replace this line by:
`if named_nodes_parents[params[param]]:`
If `params[param]` is not a named node, then I don't see how we could decide whether it should be added to givens or not. Maybe it could be added just to be safe.